keps/sig-scheduling/20190926-default-even-pod-spreading.md

```diff
@@ -0,0 +1,278 @@
+we can consider removing `SelectorSpreadingPriority` and replacing it by an equivalent
+k8s default to the default rules for `EvenPodsSpreading`.
+
+### Risks and Mitigations
+
+`EvenPodsSpreading` has some overhead and we currently ensure that pods that don't use the
+feature get minimally affected. After default rules for `EvenPodsSpreading` is rolled out,
+all pods will run through the algorithms.
+However, we should ensure that the running overhead is not significantly higher than
+`SelectorSpreadingPriority` if using the k8s default.
+
+## Design Details
+
+### API
+
+A new structure called `TopologySpreadConstraint` is introduced to `KubeSchedulerConfiguration`.
+
+```go
+type KubeSchedulerConfiguration struct {
+	....
+	// DefaultTopologySpreadConstraints defines spreading constraints to be applied to pods
+	// that don't define any.
+	// If not specified, the scheduler applies the following default.
+	// +optional
+	DefaultTopologySpreadConstraints []TopologySpreadConstraint
+	....
+}
+
+// TopologySpreadConstraint specifies how to spread pods among the given topology.
+// Pod selectors are deduced from the resource definitions that the pod belongs to
+// (includes services, controllers, replica sets and stateful sets).
+type TopologySpreadConstraint struct {
+	MaxSkew int32
+	TopologyKey string
+	WhenUnsatisfiable corev1.UnsatisfiableConstraintAction
+}
+```
+
+Note that `TopologySpreadConstraint` is similar to `k8s.io/api/core/v1.TopologySpreadConstraint`,
+except that it doesn't define selectors.
+
+### Default rules
+
+These will be the default rules for the cluster when the operator doesn't provide any:
+
+```yaml
+defaultTopologySpreadConstraints:
+  -
+    maxSkew: 1
+    topologyKey: "kubernetes.io/hostname"
+    whenUnsatisfiable: ScheduleAnyway
+  -
+    maxSkew: 1
+    topologyKey: "failure-domain.beta.kubernetes.io/zone"
+    whenUnsatisfiable: ScheduleAnyway
+```
+
+### How user stories are addressed
+
+Let's say we have a cluster that has a topology based on physical hosts and racks.
+Then, an operator can set the following scheduler configuration:
+
+```yaml
+apiVersion: componentconfig/v1alpha1
+defaultTopologySpreadConstraints:
+  -
+    maxSkew: 5
+    topologyKey: "example.com/topology/physical_host"
+    whenUnsatisfiable: ScheduleAnyway
+  -
+    maxSkew: 15
+    topologyKey: "example.com/topology/rack"
+    whenUnsatisfiable: DoNotSchedule
+```
+
+Then, a workload author could have the following `ReplicaSet`:
+
+```yaml
+apiVersion: apps/v1
+kind: ReplicaSet
+metadata:
+  name: replicated_demo
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: demo
+  template:
+    metadata:
+      labels:
+        app: demo
+    spec:
+      containers:
+      - name: php-redis
+        image: example.com/registry/demo:latest
+```
+
+Note that the workload author didn't provide a spreading rules.
+The following rules will be applied to the pods of this replica set before running the
```

Comment:
Suggest rewording to "The following spreading rules defined in ComponentConfig will be derived and applied in runtime."

Reply:
Done